Creation of a parser for AT templates and HTML

flongo opened the pull request:
## Parser - AT-HTML grammar
### A new grammar has been created.
- It allows to parse at the same time AT templating language and HTML.
- It should parse any input string and output the AST representation. There should be no case in which the parsing fails.
- It decorates the tree with validation errors and warnings.
- It interacts with the Aria Templates framework in order to retrieve the list of statements and validate their usage.
  A stylesheet has been added for the highlighting based on the new grammar.
### Build
- build and watch npm scripts have been updated for the new grammar
### Tests utility

The test utility test/node_modules/parserTester.js has been improved. It allows to
- test failures
- execute multiple tests whose input is in the same file
- test the duration of a parsing operation (useful for performance monitoring)
- output the expected result in a file
### Tests
- test.js: Tests for the AT-HTML parser have been added.
### Node/graph utility

The following methods have been added:
- merge: merges consecutive nodes updating the location information
- addError and addWarning in order to decorate a node with validation messages
- moveLocation: moves the location of a node by a specified offset. Useful for nesting parsing operations

In order to improve performances, traverse operations have been reduced, namely the whole tree elements are no longer endowed with an id or decorated with the reference to a shared node.
### Front-end application (Web application)
- The AT-HTML mode has been set as default.
- `parse`, `fold` and `outline` actions on source change have been inhibited. Only highlight has been kept. Outline is done only upon clicking on the right tab.
- `validate` action has been introduced in order to decorate the ACE editor with warnings and messages.
- The update (synchronization of the document with the backend) is not done at every code change, but every two seconds and only if something has changed.

Close #3

ymeine commented on app/node_modules/modes/node_modules/code.js:
```diff
@@ -128,14 +130,13 @@ var Code = oop.class({
 					}
 				} catch (exception) {
+					if (this.source) {
+						fs.writeFileSync(__dirname + '/failures/fail_' + (new Date()).getTime() + '.txt', this.source, "utf-8");
```
It would be good to add the name of the parser to the _stamp_. Because knowing a parser failed parsing an input but not knowing which one is in theory useless (even if in practice we will guess which one).

Unfortunately there is nothing convenient in place to identify easily the parser. Here are the quick solutions: 
- instead of creating some parsers by **using** the `Parser` class, always **extend** the latter and provide a custom name
- otherwise add a `name`/`language` or more sophisticated [taxonomy](http://en.wikipedia.org/wiki/Taxonomy) property to the `Parser` class (and provide it at instantiation)
- trust files layouts and extract from the path of the parser module the name of the mode's folder

By the way, I don't think specifying `utf-8` is required for writing, it is used by default.

flongo replied:
I agree with you. In my mind, saving failures was just a temporary feature for internal testing. It's true that it could also be used on a real release. I suggest that we should keep it as it is right now, and add the feature to the backlog

ymeine replied:
Ok, I'll add the issue for that!

ymeine commented on app/node_modules/modes/node_modules/mode.js:
```diff
@@ -251,6 +257,49 @@ var Mode = oop.Class({
+			var errors = [], warnings = [];
+
+			if (graph == null) {
+				return [];
```
I still don't know how to handle properly both edge cases and execution errors/exceptions. But I think that if the method return a value anyway, it should at least keep it consistent. My way to do it is to always return an object.

May we return an empty object? Or an object with two empty lists?

flongo replied:
sure

ymeine commented on app/node_modules/modes/node_modules/mode.js:
```diff
@@ -251,6 +257,49 @@ var Mode = oop.Class({
+		/**
+		 * Traverses the graph to look for errors and warnings
+		 * @param {Code} code
+		 * @return {Object} it contains the list od errors and warnings
```
Small typo: _od_ instead of _of_

ymeine commented on app/node_modules/modes/node_modules/node.js:
```diff
@@ -221,6 +208,34 @@ var Node = oop.Class({
+		 * @return {Node}
+		 */
+		merge: function (nodes) {
+			if (!nodes || nodes.length === 0) {
```
For consistency throughout the whole source code please use `nodes == null` instead of `!node`.

ymeine commented on app/node_modules/modes/node_modules/node.js:
```diff
@@ -221,6 +208,34 @@ var Node = oop.Class({
+		 */
+		merge: function (nodes) {
+			if (!nodes || nodes.length === 0) {
+				return nodes;
```
I think we should rather throw an error here. Still the errors/exceptions management in this application is not in place, but we should give: 
- a message
- a copy of the actual input

ymeine commented on app/node_modules/modes/node_modules/node.js:
```diff
@@ -221,6 +208,34 @@ var Node = oop.Class({
+			if (!nodes || nodes.length === 0) {
+				return nodes;
+			}
+			var newNodeObject = {
```
I would call it `nodeSpec` (this is my code style and I think `spec` is something I use often). It's the object specifying how to build a new node

However, the current name is fine too, I'm open to suggestions and also to wild dismissal :wink:

flongo replied:
`spec` stands for?

ymeine replied:
_Specifications_. It's the specifications of the new object. This is how I name the raw objects used for constructor inputs.

ymeine commented on app/node_modules/modes/node_modules/node.js:
```diff
@@ -409,6 +399,50 @@ var Node = oop.Class({
+		},
+
+		/**
+		 * Shifts the lcoation property of a certain offset
```
Small typo: _lcoation_ instead of _location_

ymeine commented on app/node_modules/modes/editor.js:
```diff
@@ -8,7 +8,11 @@ modeManager.add([
 		mode: require('./html')
-	},
+	}, {
+		names: ['athtml'],
```
Maybe you could directly add the aliases for this mode, it doesn't cost much (a simple bunch of pointers). I would essentially add `at-html`, but also: 
- `at_html`
- `at html`

And full upper case variants like `ATHTML`, `AT-HTML`, ... (we won't handle any typo anyhow, the method to handle aliases should rather change in this case)

ymeine commented on public/actions.js:
Either we also create a local variable for `htmlInfo`, or we don't for this one. Just we should keep it consistent, since they are used the same way.

ymeine commented on public/actions.js:
To avoid code duplication, I would create a simple function taking a list of messages and the type of the messages, and then it would create the annotations for you by iterating over the list and setting the proper type.

(well, I know that anyway the code of the Web application is a mess and would need a lot of refactoring!)

ymeine commented on public/index.js:
```diff
@@ -51,6 +51,10 @@ tabs.add({id: 'folding-data', label: 'Folding (data)'});
 tabs.add({id: 'ast-data', label: 'AST (data)'});
 tabs.add({id: 'outline-data', label: 'Outline (data)'});
 
+// Actions on Tabs --------------------------------------------------------------
```
What is that section for?

flongo commented on public/actions.js:
Sorry, I don't understand...

flongo commented on public/actions.js:
Good idea!

ymeine commented on app/node_modules/modes/at-html/parser/README.md:
```diff
@@ -0,0 +1,25 @@
+* `grammar.js`: generated from `grammar.pegjs`, it is the actual parser
+* [`grammar.pegjs`](./grammar.pegjs): grammar of the parser
+* [`options.json`](./options.json): options used for the generation of the parser
+* [`statements.js`](./statements.js): Module interacting with the parser in order to validate and track AT statements and also HTML tags. It retrieves information from ariatemplates source code in order to validate statements.
```
I would rather say: 

_It retrieves information using the Aria Templates framework_

instead of: 

_It retrieves information from ariatemplates source code_

For me, said as it is currently, it means you are using directly the source of the framework, not what it does.

ymeine commented on app/node_modules/modes/at-html/index.js:
```diff
@@ -0,0 +1,17 @@
+
+	stylesheet: require('./stylesheet'),
+	highlightDefaultStyleNodes: [],
+	outline: require('./outline')
```
Actually this is not used for now, this is experimental (and this is not a priority). This was copied from the other modes, but for THIS mode you should leave everything that is useless for now. Let's have a clean stuff for this one :wink:

flongo replied:
I will remove this line and the `outline.js` file

ymeine commented on app/node_modules/modes/at-html/outline.js:
```diff
@@ -0,0 +1,15 @@
+var factory = require('outline/ruleset').factory;
```
Considering my comment in `app/node_modules/modes/at-html/index.js`, you can delete this file.

ymeine commented on public/actions.js:
Sorry, not `htmlInfo`, but `html`.

ymeine commented on app/node_modules/modes/at-html/stylesheet.js:
```diff
@@ -0,0 +1,62 @@
+
+	styles: {
+		'block-statement': {
+			color: new Color([150, 150, 0])
```
I know I didn't document it, but the factory will take care of calling the `Color` constructor using the value you provided for the `color` property.

It could be simplified like this: `{color: [150, 150, 0]}`

ymeine commented on app/node_modules/modes/at-html/stylesheet.js:
```diff
@@ -0,0 +1,62 @@
+			color: Color.greyScale(100)
+		},
+		'param': {
+			color: new Color('blue')
```
Same as above, you can directly pass `{color: 'blue'}`

In this case (still it's not documented unfortunately) you can even pass `'blue'` directly to the style, like this: `'param': 'blue'`. The factory will understand you defined only one property of the style, by default the color. However, it's maybe too implicit in some cases.

ymeine commented on app/node_modules/modes/at-html/parser/statements.js:
```diff
@@ -0,0 +1,366 @@
+			 * @param {String} stm Statement name
+			 * @return {Boolean}
+			 */
+			isRoot: function(stm) {
```
I would say `return stm == "Template"` is enough.

Because with these tests you do that: 
- first exclude non-existing statements, that is equivalent to non-valid names
- then exclude all statement names that correspond to widgets
- finally exclude all statement names that are not `"Template"`

You can keep the first test if you want to specifically return `null` if it is not a valid statement, but I would leave the responsibility to the user to do the check himself. Otherwise, add to the JSDoc the fact this method can return `null`.

flongo replied:
I agree, I'll remove previous checks.

ymeine commented on app/node_modules/modes/at-html/parser/statements.js:
```diff
@@ -0,0 +1,366 @@
+			 */
+			getType: function(stm) {
+				if (!stm) {
+					return "statement";
```
Why in this case would a `null` input turn out to be a `statement`? Isn't it _not valid_?

flongo replied:
I use the `getType` method in order to create error messages like "Invalid statement name" for an input like this "{".

So the `getType` is not about being valid or not, but it is more on whether we are talking about a statement or a widget.

If you haven'r typed anything yet, I am assuming you want to type a statement.

ymeine commented on app/node_modules/modes/at-html/parser/statements.js:
```diff
@@ -0,0 +1,366 @@
+			},
+
+			/**
+			 * Returns true if the statement can be in a Macro, false if it cannot, undefined or null if it can be used both inside and outside of a macro
```
According to the end of this sentence, the name of the method, and the choice to use `false`, `true` and `null`, I guess you mean: 

"Returns true if the statement **must** be in a Macro, false if it cannot, ..."

flongo replied:
Indeed, I do

ymeine commented on app/node_modules/modes/at-html/parser/statements.js:
```diff
@@ -0,0 +1,366 @@
+			 */
+			isInMacro: function(stm) {
+				if (!this.isValid(stm)) {
+					return null;
```
I don't really like the fact to use `null` in addition to `true` and `false` to represent the result of something that can have three states. The semantics behind is not well defined. Moreover, doesn't it conflict with that in this case? Here `null` means it's not valid, not that there's no constraint towards macros. Anyway in this case I would still leave the responsibility to the user for that check. Or otherwise throw an exception.

flongo replied:
If it is `undefined` or `null`, it means that it can be used either in a macro or outside. That's the way the description is done in the class of the framework that lists the available statements.

I might change it to return `"inside"`, `"outside"` or `"any"`, but I kind of like the usage of `null`, `true` and `false` for three states.

ymeine commented on app/node_modules/modes/at-html/parser/statements.js:
```diff
@@ -0,0 +1,366 @@
+			/**
+			 * Validates the opening tag of a block AT statement, an inline AT statement, the opening tag of an HTML element or an inline HTML element
+			 *
+			 * @param {Node} node Node containing the opening tag of a block statement/element or an inline statement/tag
```
I guess you mean "or an inline statement/**element**" (to be consistent)

ymeine commented on app/node_modules/modes/node_modules/mode.js:
```diff
@@ -251,6 +257,52 @@ var Mode = oop.Class({
+			var graph = code.graph;
+			var errors = [], warnings = [];
+
+			if (graph == null) {
```
It's better to set the two previous references to empty arrays, and wrap the code below - `return` statement excluded - in the `else`. This way, if one day someone wants to change the name of the keys of the returned object (this is what makes the interface), he will have to do it only at one place, it's less error prone.

ymeine commented on test/node_modules/parserTester.js:
```diff
@@ -1,40 +1,67 @@
-		output: actualOutput,
-		result: objectIncludes(params.output, actualOutput)
-	};
+	var start = now(), end;
```
Makes more sense to me not to assign this reference now, but only before the operation you want to measure, that is before `parser.parse` below. This is clearer both for variables definition (since there are two on one lines), and for the goal of these.

flongo replied:
I have moved it one line down. I think it is enough.

ymeine commented on test/node_modules/parserTester.js:
```diff
@@ -1,40 +1,67 @@
+		end = now();
+		var actualOutput = convertToObject(parsedTree);
+
+		return {
```
I would prefer pulling those return statements out merged in a single one. In the try/catch you would just set get the failure state in a boolean variable.

Outside, you can compute all the rest. This way you would avoid repeating these pieces of code: 
- keys of the object
- `end = now()`
- `duration: end - start`

This would give something like: 

``` javascript
return {
    output: failure ? null : convertToObject(parsedTree),
    result: expectedFailure ? failure : (failure ? false : objectIncludes(params.output, actualOutput)),
    duration: end - start,
    failure: failure
};
```

(feel free and advised to split the nested ternary expressions)

flongo replied:
Done!

ymeine commented on test/node_modules/parserTester.js:
```diff
@@ -1,40 +1,67 @@
+ * - one and two are equal strings
  * - all properties contained in one (apart from children) are also cotained in two, and they are equal
  * - one and two have the same number of children
  * - each child of one is included in the corresponding child of two
```
Please quote the words referencing code, like `one` or `two`, I was confused at first reading because of the closeness with English ^^ (after talking about _one_, I would have expected talking about _the other_ in natural language, not _two_)

ymeine commented on test/node_modules/parserTester.js:
```diff
@@ -1,40 +1,67 @@
  * @param {Object} two
  * @return {Boolean}
  */
 var objectIncludes = function(one, two) {
```
Maybe better rename the parameters: 
- `one` :arrow_right: `set`
- `two` :arrow_right: `subset`

(maybe find a better name than `set`, but the idea is here)

flongo replied:
one -> small, two -> big

ymeine replied:
Better. It shows me I confused the two: `two` is the set and `one` the subset.

ymeine commented on test/node_modules/parserTester.js:
```diff
@@ -1,40 +1,67 @@
+ *
+ * If the two inputs are objects, it returns true if
+ * - one and two are equal strings
  * - all properties contained in one (apart from children) are also cotained in two, and they are equal
```
Typo: _cotained_ instead of _contained_

ymeine commented on test/test.js:
```diff
@@ -1,65 +1,440 @@
-	 *     - name (mandatory): a string with the name of the test. It is important in order to state where to find the input string and the expected type
-	 *     - rule: the name of the rule to test
-	 *     - multiple: see the parserTester class for more expalantion
+	 * Returns the test function for the correctedness of the result. Important to define it like this in order to avoid defining its closure context in the following for loop
```
Typo: _correctedness_ instead of _correctness_

ymeine commented on test/test.js:
```diff
@@ -1,65 +1,440 @@
+		];
+
+		runTests(tester, tests);
   });
```
Fix the indentation.

flongo committed:
Creation of a parser for AT templates and HTML

Parser - AT-HTML grammar
A new grammar has been created.
* It allows to parse at the same time AT templating language and HTML.
* It should parse any input string and output the AST representation. There should be no case in which the parsing fails.
* It decorates the tree with validation errors and warnings.
* It interacts with aria templates source code in order to retrieve the list of statements and validate their usage.
A stylesheet has been added for the highlight based on the new grammar.


Build
* build and watch npm scripts have been updated for the new grammar


Tests utility
The test utility test/node_modules/parserTester.js has been improved. It allows to
* test failures
* execute multiple tests whose input is in the same file
* test the duration of a parsing operation (useful for performance monitoring)
* output the expected result in a file


Tests
* test.js: Tests for the AT-HTML parser have been added.


Node/graph utility
The following methods have been added:
* merge: merges consecutive nodes updating the location information
* addError and addWarning in order to decorate a node with validation messages
* moveLocation: moves the location of a node by a specified offset. Useful for nesting parsing operations
In order to improve performances, traverse operations have been reduced, namely the whole tree elements are no longer endowed with an id or decorated with the reference to a shared node.


Front-end application
* The AT-HTML mode has been set as default.
* Parse, fold and outline actions on source change have been inhibited. Only highlight has been kept. Outiline is done only upon clicking on the right tab.
* Validate action has been introduced in order to decorate the ACE editor with warnings and messages.
* The update is niot done at every code change, but every two seconds and only if something has changed.

Close #3

flongo commented:
Integrated in 8242a22ca37a5847b6ab78eef012d1c1596bb5b1
